architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+
+#### Resource Types
+
+The resource type registration APIs add two new resource types:
+
+- `System.Resources/resourceProviders` for resource providers.
+- `System.Resources/resourceProviders/resourceTypes` for resource types.
+
+These new resource types define the full CRUDL set of APIs. These APIs are optimized for operations that need to modify or deeply understand resource types. 
+
+`resourceTypes` is our first nested resource type in UCP/Radius. This is a good usecase for a nested resource types because:
+
+- The `resourceTypes` body will be be large due to the inclusion of an OpenAPI schema. 
+- Users need granular registration and unregistration of individual resource types. (See scenario 2 above).
+
+---
+
+These new resource types are scoped to the Radius plane. 
+
+> eg: `/planes/radius/local/providers/System.Resources/resourceProviders`
+
+The rationale for this is to support registration/unregistration per-tenant of Radius rather than per-installation.
+
+#### Resource Type Queries
+
+The proposal also defines two new **non-resource** APIs:
+
+- `GET /planes/radius/{plane}/providers/`: list all resource providers.
+- `GET /planes/radius/{plane}/providers/{provider}` shows a specific resource provider.
+
+These APIs are optimized for queries and are consistent with similar APIs in ARM. To understand why this is important, remember that the OpenAPI schema for a resource type can be quite large. Some functionality in Radius will need to query resource types frequently, consider the need for the `rad recipe register` to validate a resource type. Building a separate API for queries allows us to optimize the behavior.
+
+These APIs can also be implemented consistently across all planes in the future. For example Azure doesn't support user-defined types, but it does support looking up the set of resource types.
+
+Lastly, there will still be some system-defined types. For example `System.Resources/resourceProviders` is a system defined type. The query API can return these, but the `System.Resources/resourceProviders` cannot. Imagine for a minute what would happen if users could remove the capability to specify a type.
+
+**Open Question:** Is *query* the right term for this? The term will show up in docs and in the TypeSpec + generated client + our code. Let's pick the term that's least confusing to us. I'm open to suggestions but I'll just roll with *query* unless there's a better idea.
+
+#### Locations
+
+UCP is a multi-location control-plane like the control-plane that inspired it (ARM). We don't use UCP/Radius in any multi-location scenarios right now but it's important we don't design APIs that lock us into a single-region design. ARM has a mature multi-location design but doesn't expose to users the ability to register types. However we can learn a lot from the scenarios supported by ARM for rolling out new API versions and new types.
```

Comment:
In addition to location, ARM allows resources/providers to be registered to specific subscriptions. Is there an analogy for that in Radius? For example, would a user context be applied in which specific resource types/providers are only available to a specific user? 

A valid scenario for this would be white glove onboarding of a specific team to a new resource type, or a controlled rollout of a new version of a type in which the operator wants to control who has access to the type and provider.

Reply:
Great questions.

This is covered in the alternatives section below. My opinion  is that I *hope* this complexity isn't needed.

- Azure doesn't have the ability to define your own resource providers or types, so the registration functionality is needed.
- Radius/UCP does have the ability to define your own resource types so the registration functionality would be redundant. 

---

> A valid scenario for this would be white glove onboarding of a specific team to a new resource type

In the Azure context is this is a combination of the registration feature + feature flagging. 

---

The registration feature provides a "callback" to the resource provider so it can initialize itself (mostly identities) and do checks it needs to do. 


If we need a "callback" in UCP/Radius in the future we could add it, and we'd tie to to resource provider creation rather than registration.